Extend process-wide machinery to handle more idioms (#32)
bdaiinstitute/ros_utilities: README.md

# ros_utilities
Wrappers and other utilities for ROS2

## Getting started

### Process-wide APIs

These APIs wrap [`rclpy`](https://github.com/ros2/rclpy) APIs to provide a simpler UX in common use cases.
All APIs are thread-safe and exception-safe. Resource management is automatic. By default, callbacks are
dispatched in the background, enabling both synchronous and asynchronous programming.

These APIs are also quite handy to reconcile past [`rospy`](http://wiki.ros.org/rospy) experience with ROS 2.

#### Setting up single node processes

Just decorate your executable entrypoint (or `main` function) with `bdai_ros2_wrappers.process.main`:

```python
import logging
import time

import bdai_ros2_wrappers.process as ros_process

import std_msgs.msg

@ros_process.main()

> def entrypoint():

Reviewer: The examples and explanations look great.

Going to ask for type hints just to keep it consistent with our code expectations.

# ros_utilities
Wrappers and other utilities for ROS2

> ## Getting started

Reviewer 2: I think this should also point the reader to the examples directory

### Process-wide APIs

These APIs wrap [`rclpy`](https://github.com/ros2/rclpy) APIs to provide a simpler UX in common use cases.
All APIs are thread-safe and exception-safe. Resource management is automatic. By default, callbacks are
dispatched in the background, enabling both synchronous and asynchronous programming.

> These APIs are also quite handy to reconcile past [`rospy`](http://wiki.ros.org/rospy) experience with ROS 2.

Reviewer 2: I feel like a little bit more conceptual stuff here might help with understanding the examples.  I'd throw in some high level stuff about how the backend works.  For instance, a "process-wide node" is referenced a few times - I think it'd be worth saying what that is.

Author: Good idea. Done in https://github.com/bdaiinstitute/ros_utilities/pull/32/commits/efb0db0d1ef57bdecc49aaff0dcbe9222defb1e3.

#### Setting up single node processes

> Just decorate your executable entrypoint (or `main` function) with `bdai_ros2_wrappers.process.main`:

Reviewer 2: ```suggestion
To create simple single node processes that can block in common sense ways similar to ROS1's `rospy`, just decorate your executable entrypoint (or `main` function) with `bdai_ros2_wrappers.process.main`:
```

Reviewer 2: Can you also specify what the person should see if they run this program?  I actually don't know what `create_task` does...

    pub = node.create_publisher(std_msgs.msg.String, "test", 1)

    def callback() -> None:
        time.sleep(10)  # you can block in a callback
        pub.publish(std_msgs.msg.String(data="testing"))
    executor = ros_process.executor()  # or entrypoint.executor
    assert executor is not None
    executor.create_task(callback)

    time.sleep(10)  # you can block in the main thread

    node.get_logger().info("testing")
    logging.info("testing")  # you can use Python logging

    return  # no need to cleanup or shutdown, it is automatic

if __name__ == "__main__":
    entrypoint()
```

This is ideal for quick prototyping and simple scripts.

#### Setting up multi-node processes

You can spin as many ROS 2 nodes as you need, and skip the default process-wide node if unnecessary:

```python
import logging
import time

from typing import Any, List

import bdai_ros2_wrappers.process as ros_process
from bdai_ros2_wrappers.node import Node

def graph(**kwargs: Any) -> List[Node]:
    # make sure to forward all keyword arguments for proper loading!
    return [Node("my_node", **kwargs), Node("my_other_node", **kwargs)]

@ros_process.main(prebaked=False)
def main() -> None:

>     ros_process.load(Node, "_hidden_node")  # or main.load

Reviewer 3: Is this to show that we can throw in another node into the processes and that when we call spin(graph) will that then spin all 3 nodes?

Author: Exactly. And that can be done from anywhere below `main` in the call stack.

Author: It will use defaults, but you can configure it before that happens:

```python
from bdai_ros2_wrappers.executors import AutoScalingMultiThreadedExecutor
from bdai_ros2_wrappers.executors import foreground

# ...

@ros_process.main(prebaked=False)
def main() -> None:
    with foreground(AutoScalingMultiThreadedExecutor(max_threads=100)) as main.executor:
        ros_process.load(Node, "_hidden_node")  # or main.load
        ros_process.spin(graph)  # or main.spin
```

Reviewer 3: Awesome!

Reviewer 2: Can the `README` be updated to explain that this will spin all three nodes and perhaps address a little more what `Load` does?  I had the same question reading this as @mpickett-bdai

Author: Done in https://github.com/bdaiinstitute/ros_utilities/pull/32/commits/efb0db0d1ef57bdecc49aaff0dcbe9222defb1e3.

    ros_process.spin(graph)  # or main.spin, and it will block!

if __name__ == "__main__":
    main()
```

Note the use of ROS 2 node(s) factories, to load (or spin) entire collections at once.

#### Setting up interactive multi-node applications

> For interactive applications, you will most likely want to keep automatic spinning in place:

Reviewer 2: ```suggestion
For interactive applications, you will most likely want to keep automatic spinning in place.  Here's an example that waits for user input, sleeps for 5 seconds, and then echoes that user input from a callback without deadlocking:
```

```python
import logging
import time

from typing import Any, List

from bdai_ros2_wrappers.futures import wait_for_future
from bdai_ros2_wrappers.node import Node
import bdai_ros2_wrappers.process as ros_process

def graph(**kwargs: Any) -> List[Node]:
    # make sure to forward all keyword arguments for proper loading!
    return [Node("my_node", **kwargs), Node("my_other_node", **kwargs)]

def work(prompt: str) -> None:
    def worker() -> None:
        node = ros_process.node()
        assert node is not None
        time.sleep(5.0)  # sleeping is safe!
        node.get_logger().info(prompt + " done!")
    executor = ros_process.executor()
    assert executor is not None
    future = executor.create_task(worker)  # dispatch worker to executor
    wait_for_future(future)  # block until worker is done

@ros_process.main()
def main() -> None:
    ros_process.load(graph)  # or main.load
    while True:
        work(input())

if __name__ == "__main__":
    main()
```

> #### Setting up command-line single node applications

Reviewer 2: Could there be a sentence here about what this is showing and also what `managed` means?

import logging
import time

from typing import Any

from bdai_ros2_wrappers.futures import wait_for_future
from bdai_ros2_wrappers.node import Node
import bdai_ros2_wrappers.process as ros_process

class Application(Node):

    def __init__(self, robot_name: str, **kwargs: Any) -> None:
        super().__init__("app_node", **kwargs)
        self.robot_name = robot_name

def cli() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser()
    parser.add_argument("robot_name")
    parser.add_argument("-v", "--verbose", action="store_true")
    # define ROS 2 aware process arguments based on CLI arguments
    parser.set_default(process_args=lambda args: dict(
        prebaked=False, forward_logging=args.verbose))
    return parser

@ros_process.main(cli())
def main(args: argparse.Namespace) -> None:
    # set process-wide node for log forwarding to take effect
    with ros_process.managed(Application, args.robot_name) as main.node:
        if args.verbose:
            root = logging.getLogger()
            root.setLevel(logging.INFO)
        logging.info("Application started!")
        ros_process.spin()  # or main.spin

if __name__ == "__main__":
    main()
```

> Note that process arguments are set via CLI. Also, note that a process-wide node is set explicitly (rather than implicitly). 

Reviewer 2: In what cases does fine grained control over ROS 2 configuration matter?  Could you expand a little here on why you want to set a process-wide node explicitly?

Author: Clarified in https://github.com/bdaiinstitute/ros_utilities/pull/32/commits/efb0db0d1ef57bdecc49aaff0dcbe9222defb1e3.

The same can be done for process-wide executors. This allows fine grained control over ROS 2 configuration.

#### Writing ROS 2 aware libraries

You can use process-wide APIs to fetch ROS 2 defaults deep down call hierarchies, when passing nodes explicitly adds unnecessary clutter:

```python
import time

import bdai_ros2_wrappers.process as ros_process
import bdai_ros2_wrappers.scope as ros_scope
from bdai_ros2_wrappers.node import Node

class SomeFeature:

    def __init__(self) -> None:
        self.node = ros_scope.load(Node, "my_node")

    def shutdown(self) -> None:
        ros_scope.unload(self.node)  # or leave it behind until process ends

@ros_process.main()
def main() -> None:
    feature = SomeFeature()
    try:
        time.sleep(10.0)
    finally:
        feature.shutdown()

if __name__ == "__main__":
    main()
```

> Note the use of `ros_scope` instead of `ros_process`. Both offer the same APIs, but the former reaches out to the innermost scope. This allows the library to work in more complex, multi-threaded, multi-scope applications, and for testing cases.

Reviewer 2: I'm still not clear at this point on when I'd use `ros_scope` vs `ros_process`.  Could you give an example of when scope is needed?  Is the above such an example?  What would happen if process was used instead?

Author: Should be clearer now in https://github.com/bdaiinstitute/ros_utilities/pull/32/commits/efb0db0d1ef57bdecc49aaff0dcbe9222defb1e3, as the notion of scope is introduced and explained earlier.

#### Writing ROS 2 aware tests

> To write tests, setup a global ROS 2 aware scope instead of a process.

Reviewer 2: ```suggestion
To write tests, set up a global ROS 2 aware scope instead of a process.
```

Author: Fixed in https://github.com/bdaiinstitute/ros_utilities/pull/32/commits/efb0db0d1ef57bdecc49aaff0dcbe9222defb1e3.

Consider using namespaces to avoid default, hidden ROS 2 names.

Using [`pytest`](https://docs.pytest.org/en/6.2.x/contents.html) (recommended):

```python
from typing import Iterable

import bdai_ros2_wrappers.scope as ros_scope
from bdai_ros2_wrappers.scope import ROSAwareScope

import pytest

@pytest.fixture
def ros() -> Iterable[ROSAwareScope]:
    with ros_scope.top(global_=True, namespace="fixture") as top:

>         yield top

Reviewer 2: It's probably worth explaining what this function does for folks who aren't super up on pytest fixtures and yield.

Author: Added a docstring in https://github.com/bdaiinstitute/ros_utilities/pull/32/commits/efb0db0d1ef57bdecc49aaff0dcbe9222defb1e3.

#### Setting up interactive multi-node applications

For interactive applications, you will most likely want to keep automatic spinning in place. Here's an example that waits
for user input, sleeps for 5 seconds, and then echoes that user input from a callback without deadlocking:

```python
import logging
import time

from typing import Any, List

from bdai_ros2_wrappers.futures import wait_for_future
from bdai_ros2_wrappers.node import Node
import bdai_ros2_wrappers.process as ros_process

def graph(**kwargs: Any) -> List[Node]:
    # make sure to forward all keyword arguments for proper loading!
    return [Node("my_node", **kwargs), Node("my_other_node", **kwargs)]

def work(prompt: str) -> None:
    def worker() -> None:
        node = ros_process.node()
        assert node is not None
        time.sleep(5.0)  # sleeping is safe!
        node.get_logger().info(prompt + " done!")
    executor = ros_process.executor()
    assert executor is not None
    future = executor.create_task(worker)  # dispatch worker to executor
    wait_for_future(future)  # block until worker is done

@ros_process.main()
def main() -> None:
    ros_process.load(graph)  # or main.load
    while True:

>         work(input())

Reviewer: Can you use `input` if this were launched via a launch file? My current understanding after digging through launch is you cannot get stdin from the launch process into a node?

Author: Right, you cannot. `launch` does not have provisions to pipe its own `stdin` to that of a process.

Reviewer: Ok, can mention that here? I would be concerned that someone uses `input` based on this example and then later has trouble launching.

Author: Done in https://github.com/bdaiinstitute/ros_utilities/pull/32/commits/1c1390b1712dee796beef2905f64f27a3e92b56a.
